### PR TITLE
Rigorously define equals method on Term and Quad

### LINKED ISSUE #133: Quad interface: when does equals return true?
The spec says:

> equals() returns true if and only if the argument is a) of the same type b) has all components equal.

The only parameter is of type `Quad`, therefore (a) is always right, isn't it? Or what is meant with "of the same type"?

If my assumption is right, i would suggest removing the (a) part from the text.

(ref #130)

### PR DESCRIPTION
**Pull request by RubenVerborgh**
Closes #133.

### CODE REVIEW

File: interface-spec.html [hunk 1]
```diff
@@ -146,8 +146,18 @@ <h3><dfn>Term</dfn> interface</h3>
...
-      <code>termType</code> and the same contents (as defined by concrete subclasses).
+      <dfn id="term-equals">equals()</dfn> returns <code>true</code>
+      when called with parameter <code>other</code>
+      on an object <code>term</code> if and only if
+      all of the conditions below hold:
+    </p>
+    <ul>
+      <li><code>other</code> is <em>neither</em> <code>null</code> nor <code>undefined</code>;</li>
+      <li><code>term.termType</code> is the same string as <code>other.termType</code>;</li>
+      <li><code>other</code> follows the additional constraints followed by the most specific <code>Term</code> interface implemented by <code>term</code> (as detailed below);</li>
```
Comment (rubensworks) [inline interface-spec.html]: `followed by the most specific <code>Term</code>`

This text makes it seem that multiple interfaces could be implemented, which shouldn't be possible. Should this be _"the more specific Term"_ instead?

File: interface-spec.html [hunk 2]
```diff
@@ -233,9 +247,12 @@ <h3><dfn>Literal</dfn> interface</h3>
...
     </p>
     <p>
-      <dfn>equals()</dfn> returns <code>true</code> if and only if <code>other</code> has
-      <code>termType</code> <code>"Literal"</code> and the same <code>value</code>,
-      <code>language</code>, and <code>datatype</code>.
+      <dfn>equals()</dfn> returns <code>true</code> if and only if
+      <a href="#term-equals">all general <code>Term#equals</code> conditions</a> hold,
+      <code>term.value</code> is the same string as <code>other.value</code>,
+      <code>term.language</code> is the same string as <code>other.language</code>, and
+      <code>term.dataType.equals(other.dataType)</code> evaluates to a truthy value;
```
Comment (rubensworks) [inline interface-spec.html]: `<code>term.datatype.equals(other.datatype)</code>`

File: interface-spec.html (hunk 1 above)
Comment (blake-regalia) [reply interface-spec.html]: I must admit that I don't understand what this last item means about the more specific Term interface?

File: interface-spec.html [hunk 3]
```diff
@@ -334,8 +354,20 @@ <h3><dfn>Quad</dfn> interface</h3>
...
+      when called with parameter <code>other</code>
+      on an object <code>quad</code> if and only if
+      all of the conditions below hold:
+    </p>
+    <ul>
+      <li><code>other</code> is <em>neither</em> <code>null</code> nor <code>undefined</code>;</li>
+      <li><code>quad.subject.equals(other.subject)</code> evaluates to a truthy value;</li>
+      <li><code>quad.predicate.equals(other.predicate)</code> evaluates to a truthy value;</li>
+      <li><code>quad.object.equals(other.object)</code> evaluates to a truthy value;</li>
+      <li><code>quad.graph.equals(other.graph)</code> evaluates to a truthy value;</li>
```
[the remark this answers is not in the document]
Comment (RubenVerborgh) [reply interface-spec.html]: The idea was to relax, so you don’t have to write `=== true`; you might not own the implementations of the component methods (terms could be imported).

File: interface-spec.html (hunk 1 above)
Comment (RubenVerborgh) [reply interface-spec.html]: It means NamedNode/Literal/… will rephrase, probably enumerate.

File: interface-spec.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (blake-regalia) [reply interface-spec.html]: Okay, so then we also have to check that `quad.[subject/predicate/object/graph].equals` are indeed functions before using them yes? Otherwise they will throw an error.

**Comment by RubenVerborgh** [comment]
Checking for function is extra work; checking for truthy instead of true is less. I don’t mind though; let me know if you want true and I’ll change, I just don’t think it’s necessary to go the extra mile.

**Comment by blake-regalia** [comment]
Yes, but isn't the idea to always return `true` or `false` and not throw an error? What if i pass a term to quad constructor such that `quad.term.equals === null`? We need to test at runtime that `.equals` is a function for each component of quad; and that this is another consequence of #132 / #133

**Comment by RubenVerborgh** [comment]
I don’t feel strongly, will just write true.

File: interface-spec.html [hunk 4]
```diff
@@ -146,12 +146,24 @@ <h3><dfn>Term</dfn> interface</h3>
...
+      <dfn id="term-equals">equals()</dfn> returns <code>true</code>
+      when called with parameter <code>other</code>
+      on an object <code>term</code> if and only if
+      all of the conditions below hold:
+    </p>
+    <ul>
+      <li><code>other</code> is <em>neither</em> <code>null</code> nor <code>undefined</code>;</li>
+      <li><code>term.termType</code> is the same string as <code>other.termType</code>;</li>
+      <li><code>other</code> follows the additional constraints of the specific <code>Term</code> interface implemented by <code>term</code>
+        (e.g., <a href="#namednode-interface"><code>NamedNode</code></a>,
```
Comment (bergos) [inline interface-spec.html]: Just `<a>NamedNode</a>` is enough to link internal. Respec detects that it's an interface and will format it like `code`.

File: interface-spec.html [hunk 5]
```diff
@@ -146,12 +146,24 @@ <h3><dfn>Term</dfn> interface</h3>
...
+        <a href="#literal-interface"><code>Literal</code></a>, …);</li>
+    </ul>
+    <p>
+      otherwise, it returns <code>false</code>.
     </p>
   </section>
 
-  <section data-dfn-for="NamedNode">
+  <section id="namednode-interface" data-dfn-for="NamedNode">
```
Comment (bergos) [inline interface-spec.html]: Using the simple version of `<a></a>` doesn't require any `id` attributes.

File: interface-spec.html [hunk 6]
```diff
@@ -169,12 +181,14 @@ <h3><dfn>NamedNode</dfn> interface</h3>
       <dfn>value</dfn> the IRI of the named node (example: <code>"http://example.org/resource"</code>).
     </p>
     <p>
-      <dfn>equals()</dfn> returns <code>true</code> if and only if <code>other</code> has
-      <code>termType</code> <code>"NamedNode"</code> and the same <code></code>value</code>.
+      <dfn>equals()</dfn> returns <code>true</code> if and only if
+      <a href="#term-equals">all general <code>Term#equals</code> conditions</a> hold
```
Comment (bergos) [inline interface-spec.html]: Respec would also handle this case if `<a>Term.equals</a>` is used. As the link would not cover the current text, maybe change it to:

`if and only if all general condititions defined in <a>Term.equals</a>`

**Comment by blake-regalia** [comment]
There is still an outstanding edge case that is not covered by the algorithm which I brought up in my last comment; namely that `.equals` needs to be tested for otherwise what happens when:

```js
factory.quad({
	termType: 'NamedNode',
	value: 'http://ex.org/s',
}, {
	termType: 'NamedNode',
	value: 'http://ex.org/p',
}, {
	termType: 'NamedNode',
	value: 'http://ex.org/o',
}, {
	termType: 'NamedNode',
	value: 'http://ex.org/g',
}).equals(other);
```

will throw: `Uncaught TypeError: {...}.equals is not a function`

**Comment by RubenVerborgh** [comment]
@blake-regalia We are assuming now that they are valid `Term`s; I can state that assumption somewhere explicitly, but I'd prefer to keep it. Too complex and expensive otherwise.

Also see #137.

File: interface-spec.html (hunk 6 above)
Comment (RubenVerborgh) [reply interface-spec.html]: Mmm, don't personally like this, but WebIDL has to be followed, thanks for noticing.

**Comment by RubenVerborgh** [comment]
@bergos' comments processed, pending @blake-regalia response.

**Comment by blake-regalia** [comment]
> We are assuming now that they are valid `Terms`;
> Too complex and expensive otherwise.

@RubenVerborgh By your own logic, iff we are assuming they are valid `Terms` then we don't need to test for `null` or `undefined`! Where is the consistency in that?

**Comment by RubenVerborgh** [comment]
Null pointer exceptions are a common source of errors, I expect several magnitudes more common than malformed Term exceptions.

**Comment by blake-regalia** [comment]
> Null pointer exceptions are a common source of errors

Agreed - and I still think it would be nice to let devs know that the variable they passed to equals is actually `null` ;)  because who does that intentionally?

> I expect several magnitudes more common than malformed Term exceptions.

Well we are not just going for a majority of cases, it needs to be well-defined, consistent and fully-covered.

**Comment by RubenVerborgh** [comment]
Alright, not disagreeing, but I don't have more bandwidth to follow up on this. Re-assigning to @bergos and @blake-regalia, feel free to conclude in the way you think is most appropriate.

**Comment by elf-pavlik** [comment]
I think this relates to #104 (Nuance "plain objects (JSON) cannot be used) since @blake-regalia just used such 'plain objects' to make his case.

**Comment by bergos** [comment]
We could add to the list of conditions:

> no error is thrown when evaluating the `.equals` calls

Then we would be on the safe side.

But also think of the impact. Maybe you want to throw an error on purpose. I don't have a good example now, but maybe access control could be a use case or a native implementation with a very low level error. You don't want to swallow errors you throw on purpose. 

From my point of view it's not about catching all errors (also not null pointer exceptions which could be thrown by a `Term` implementation), it's about allowing nothing (`null`/`undefined`) or a `Quad`. `String.prototype.include` also supports my point, as the following code throws an error:

`'a'.includes({toString: () => {throw Error()}})`

**Comment by RubenVerborgh** [comment]
> > no error is thrown when evaluating the `.equals` calls
> 
> Then we would be on the safe side.

Too expensive to make such a guarantee.

**Comment by vhf** [comment]
> From my point of view it's not about catching all errors […], it's about allowing nothing (`null`/`undefined`) or a `Quad`.

Agreed.

> iff we are assuming they are valid `Terms` then we don't need to test for `null` or `undefined`!

@blake-regalia I understand this concern about consistency.

On the one hand the `.equals` algorithm suggested in this PR doesn't explicitly state that `other` should be a term but the interface does:

```
interface Term {
[…]
  boolean equals(Term other);
};
```

On the other hand the `.equals` algorithm allows for `null` and `undefined` while the interface doesn't.

What would you all think of 

```diff
interface Term {
[…]
- boolean equals(Term other);
+ boolean equals(optional Term? other);
};
```

I think it would solve this particular issue: now it makes sense that the algo has to cover `null` and `undefined`, and that `other` can only be a `Term` when defined and not null.

**Comment by elf-pavlik** [comment]
maybe @vhf you could make separate PR to the branch of this PR? Ruben mentioned having no more bandwidth left few comments above

**Comment by blake-regalia** [comment]
@vhf I think that is a great suggestion!

**Comment by blake-regalia** [comment]
@vhf In that case, I assume then it would be OK for implementations to `throw` if passed, for instance, a `string` or `number` which does not have a `.termType` property? I think then this algo would remove the criteria about checking that those properties exist at runtime, correct?

**Comment by RubenVerborgh** [comment]
> a `string` or `number` which does not have a `.termType` property?

That won't happen unless you throw it yourself; all strings and numbers have such a property.

**Comment by blake-regalia** [comment]
> That won't happen unless you throw it yourself;

To clarify, what I mean is: would throwing be allowed as opposed to returning false?

**Comment by RubenVerborgh** [comment]
I’m opposed to mandating exceptions in this low-level spec for performance reasons. We could leave the behavior undefined, worst case.

**Comment by blake-regalia** [comment]
Yes I agree and I am totally happy with leaving it undefined as long as we clarify that in the spec; i.e., this would mean that throwing or returning false are both allowed.

**Comment by vhf** [comment]
I added a commit for the first part of the latest comments: making `.equals` parameter optional and nullable.

As for the other part I'm not sure to understand your suggestion @blake-regalia , let's start here:

> In that case, I assume then it would be OK for implementations to `throw`

Why this assumption? This PR so far doesn't let you throw when you implement `.equals`, and it specifies that `other` can be undefined or null, and how to handle the case when it's undefined, null, or a Term/Quad.

**Comment by bergos** [comment]
@blake-regalia can you make a proposal what to change so the PR gets your approval?

**Comment by RubenVerborgh** [comment]
Closing this concrete proposal until we reach an agreement here: https://github.com/rdfjs/representation-task-force/issues/151

**Comment by RubenVerborgh** [comment]
Reopening based on resolution of #151.

**Comment by elf-pavlik** [comment]
@RubenVerborgh with current approvals I think you can just go ahead and merge it.

**Comment by blake-regalia** [comment]
I added @vhf for review as he had a good suggestion about the signatures.

**Comment by RubenVerborgh** [comment]
No objection to that.

**Comment by elf-pavlik** [comment]
@blake-regalia should we just apply patch @vhf posted in his comment above? https://github.com/rdfjs/data-model-spec/pull/142#issuecomment-457104605 and do manual merge?

**Comment by blake-regalia** [comment]
Okay, made the commits -- let's wait for @vhf to review and then we can merge.
